src/chapter1/06-functional-thinking.md

```diff
@@ -0,0 +1,270 @@
+    description: 'Write an awesome task app.',
+    timeLogged: 0,
+    status: 'Todo'
+  },
+  {
+    id: 3,
+    summary: 'Think of a funny joke',
+    description: 'Come up with a funny joke to lighten the mood.',
+    timeLogged: 120,
+    status: 'In progress'
+  }
+];
+```
+
+The `map` function takes one argument - a function `f` to apply to every element of the array. It returns the array resulting from applying `f` to every element of the original array.
+
+Let's say we wanted to square all the elements of `numbers`. We could write something like this:
+
+```javascript
+const result = [];
+for (const number in numbers) {
+  result.push(number ** 2);
+}
+```
+
+This is ugly and (you guessed it) *tedious*. Instead we can (and should) use the `map` function:
+
+```javascript
+const result = numbers.map(number => number ** 2);
+```
+
+The `result` array has the following content:
+
+```javascript
+[ 1, 4, 9, 16 ]
+```
+
+Consider another example. Let's say we wanted to add a long description to all the `tasks` based on the summary and the description. We can use the `map` function again:
+
+```javascript
+const longTasks = tasks.map(task => ({
+  ...task,
+  longDescription: `${task.summary}: ${task.description}`
+}));
+```
+
+> You can see why the spread syntax is so handy. Thanks to this *incredible innovation*, you only need to explicitly specify the object parts where something interesting happens.
+
+The `longTasks` array will look like this:
+
+```javascript
+[
+  {
+    id: 1,
+    summary: 'Read the MERN book',
+    description: 'Read and understand the MERN book.',
+    timeLogged: 60,
+    status: 'In progress',
+    longDescription: 'Read the MERN book: Read and understand the MERN book.'
+  },
+  {
+    id: 2,
+    summary: 'Write a task app',
+    description: 'Write an awesome task app.',
+    timeLogged: 0,
+    status: 'Todo',
+    longDescription: 'Write a task app: Write an awesome task app.'
+  },
+  {
+    id: 3,
+    summary: 'Think of a funny joke',
+    description: 'Come up with a funny joke to lighten the mood.',
+    timeLogged: 120,
+    status: 'In progress',
+    longDescription: 'Think of a funny joke: Come up with a funny joke to lighten the mood.'
+  }
+]
+```
+
+The `filter` function allows you to select elements from an array based on some condition. It takes a function `f` which returns `true` or `false` for some input(s).
+
+> A function which returns `true` or `false` is commonly referred to as a *predicate*.
+
+For example let's say we want to select all even elements from `numbers`. Here is the non-functional way:
+
+```javascript
+const result = [];
+for (const number in numbers) {
+  if (number % 2 === 0) {
+    result.push(number ** 2);
+  }
+}
+```
+
+Ugh! I don't even want to look at that. For loops and if statements all over the place. Let's rest our eyes and consider the *functional* approach:
+
+```javascript
+const result = numbers.filter(number => number % 2 === 0);
+```
+
```

Comment:
I think it would be good to point out that filter does not filter elemts out of the array but rather dictates which ones are allowed to stay, I see a lot that people reverse it